// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.controllers_grpc_interface";
option java_outer_classname = "controllers_grpc_interfaceProto";
option objc_class_prefix = "HLW";

package controllers_grpc_interface;


service GlobalToLocal{
  rpc LocalHandshake (LocalSimplifiedHandshakeRaw) returns (ACK) {}
  rpc StageHandshake (ControlOperation) returns (StageSimplifiedHandshakeRaw) {}
  rpc MarkStageReady (StageReadyRaw) returns (ACK) {}
  rpc CreateHouseKeepingRuleChannel (HousekeepingCreateChannelString) returns (ACK) {}
  rpc CreateHouseKeepingRuleObject (HousekeepingCreateObjectString) returns (ACK) {}
  rpc ExecuteHousekeepingRules (Execute) returns (ACK) {}
  rpc CreateEnforcementRule (EnforcementRuleString) returns (ACK) {}
  rpc CollectGlobalStatistics (ControlOperation) returns (StatsGlobalMap) {}
  rpc CollectEntityStatistics (ControlOperation) returns (StatsEntityMap) {}
}


message LocalSimplifiedHandshakeRaw {
  repeated string rules = 1;
}


message ControlOperation {
  int32 m_operation_id = 1; // ControlReceive identifier.
  int32 m_operation_type = 2; // Control Plane operation type (CREATE_HSK_RULE, CREATE_ENF_RULE,
  // COLLECT_STATS, ...).
  int32 m_operation_subtype = 3; // Control Plane operation subtype (HSK_CREATE_UNIT,
  // HSK_CREATE_CHANNEL, HSK_CREATE_OBJECT, ...).
  int32 m_size = 4; // Size of the RAW object to receive.
};


message StageSimplifiedHandshakeRaw {
  string m_stage_name = 1;
  string m_stage_env = 2;
  int32 m_pid = 3;
  int32 m_ppid = 4;
  string m_hostname = 5;
  string m_user = 6;
}

message StageReadyRaw {
  bool m_mark_stage = 1;
}

message HousekeepingCreateChannelString {
  string m_stage_name = 1;
  string m_stage_env = 2;
  string m_rule = 3;
}

message HousekeepingCreateObjectString {
  string m_stage_name = 1;
  string m_stage_env = 2;
  string m_rule = 3;
}

message HousekeepingCreateChannelRaw {
  uint64 m_rule_id = 1;
  int32 m_rule_type = 2;
  int64 m_channel_id = 3;
  int32 m_context_definition = 4;
  uint32 m_workflow_id = 5;
  uint32 m_operation_type = 6;
  uint32 m_operation_context = 7;
}

message HousekeepingCreateObjectRaw {
  uint64 m_rule_id = 1;
  int32 m_rule_type = 2;
  int64 m_channel_id = 3;
  int64 m_enforcement_object_id = 4;
  int32 m_context_definition = 5;
  uint32 m_operation_type = 6;
  uint32 m_operation_context = 7;
  int64 m_enforcement_object_type = 8;
  int64 m_property_first = 9;
  int64 m_property_second = 10;
}


message ACK {
  int32 m_message = 1;
}

message Execute {
  string m_stage_name = 1;
  string m_stage_env = 2;
  bool execute_all = 3;
}


message EnforcementRuleString {
  int64 m_rule_id = 1;
  string m_stage_name = 2;
  string m_operation = 3;
  map<int64, int64> env_rates = 4;
}

message EnforcementRuleRaw {
  int64 m_rule_id = 1;
  int64 m_channel_id = 2;
  int64 m_enforcement_object_id = 3;
// // int m_enforcement_object_type { 0 }; // fixme: probably not needed; remove
  int32 m_enforcement_operation = 4;
  int64 m_property_first = 5;
  int64 m_property_second = 6;
  int64 m_property_third = 7;
}

message StatsEntity {
  map<string, double> ent_stats = 1;
}

message StatsEntityMap {
  map<string, StatsEntity> stats = 1;
};

message StatsMetadata {
  double m_open_rate = 1;
  double m_close_rate = 2;
  double m_getattr_rate = 3;
  double m_metadata_total_rate = 4;
};


message StatsGlobalMap {
  repeated StatsGlobal stats = 1;
};

message StatsGlobal {
  string stage_name = 1;
  string stage_env = 2;
  double m_read_rate = 3;
  double m_write_rate = 4;
  double m_open_rate = 5;
  double m_close_rate = 6;
  double m_getattr_rate = 7;
  double m_metadata_total_rate = 8;
};


//////

service LocalToGlobal{
  rpc ConnectLocalToGlobal (ConnectRequest) returns (ConnectReply) {}
  rpc ConnectStageToGlobal (ConnectRequestStage) returns (ConnectReply) {}
}

// The request message containing the user's name.
message ConnectRequest {
  string user_address = 1;
}

// The request message containing the user's name.
message ConnectRequestStage {
  string local_address = 1;
  string stage_name = 2;
  string stage_env = 3;
  string stage_user = 4;
}

// The response message containing the greetings
message ConnectReply {
  string message = 1;
}











